hi-ml-cpath/testhisto/testhisto/utils/test_callbacks.py

```diff
@@ -0,0 +1,233 @@
+    if epochs_interval == 2:
+        assert callback.is_time_to_cache_loss_values(current_epoch) is False
+    else:
+        assert callback.is_time_to_cache_loss_values(current_epoch)
+
+    current_epoch = 5
+    assert callback.is_time_to_cache_loss_values(current_epoch)
+
+
+def test_on_train_batch_start(tmp_path: Path, mock_panda_tiles_root_dir: Path) -> None:
+    batch_size = 2
+    container = MockDeepSMILETilesPanda(tmp_path=mock_panda_tiles_root_dir)
+    container.batch_size = batch_size
+    container.setup()
+    container.create_lightning_module_and_store()
+
+    current_epoch = 5
+    trainer = MagicMock(current_epoch=current_epoch)
+    batch = next(iter(container.data_module.train_dataloader()))
+
+    callback = LossAnalysisCallback(outputs_folder=tmp_path)
+    _assert_loss_cache_contains_n_elements(callback.loss_cache, 0)
+
+    callback.on_train_batch_start(trainer, container.model, batch, 0, None)  # type: ignore
+    _assert_loss_cache_contains_n_elements(callback.loss_cache, batch_size)
+
+    batch = next(iter(container.data_module.train_dataloader()))
+    callback.on_train_batch_start(trainer, container.model, batch, 1, None)  # type: ignore
+    _assert_loss_cache_contains_n_elements(callback.loss_cache, 2 * batch_size)
+
+
+def get_loss_cache(n_slides: int = 4, rank: int = 0) -> LossDictType:
+    return {
+        ResultsKey.LOSS: list(range(1, n_slides + 1)),
+        ResultsKey.SLIDE_ID: [f"id_{i}" for i in range(rank * n_slides, (rank + 1) * n_slides)],
+        ResultsKey.TILE_ID: [f"a${i * (rank + 1)}$b" for i in range(rank * n_slides, (rank + 1) * n_slides)],
+    }
+
+
+def test_on_train_epoch_end(
+    tmp_path: Path, duplicate: bool = False, rank: int = 0, world_size: int = 1, device: str = "cpu"
+) -> None:
+    current_epoch = 5
+    n_slides_per_process = 4
+    trainer = MagicMock(current_epoch=current_epoch)
+    pl_module = MagicMock(global_rank=rank)
+
+    loss_callback = LossAnalysisCallback(outputs_folder=tmp_path, num_slides_heatmap=2, num_slides_scatter=2)
+    loss_callback.loss_cache = get_loss_cache(rank=rank, n_slides=n_slides_per_process)
+    if duplicate:
+        # Duplicate slide "id_0" and to test that the duplicates are removed
+        loss_callback.loss_cache[ResultsKey.SLIDE_ID][0] = "id_0"
+
+    _assert_loss_cache_contains_n_elements(loss_callback.loss_cache, n_slides_per_process)
+    loss_callback.on_train_epoch_end(trainer, pl_module)
+    # Loss cache is flushed after each epoch
+    _assert_loss_cache_contains_n_elements(loss_callback.loss_cache, 0)
+
+    if rank > 0:
+        time.sleep(10)  # Wait for the rank 0 to save the loss cache in a csv file
+
+    loss_cache_path = loss_callback.cache_folder / LOSS_VALUES_FILENAME.format(loss_callback.zfill_epoch(current_epoch))
+    assert loss_callback.cache_folder.exists()
+    assert loss_cache_path.exists()
+    assert loss_cache_path in loss_callback.cache_folder.iterdir()
```

Comment:
[Minor] What't the purpose of this check? Do you mean `local_cache_path.parent == loss_callback.cache_folder`?